polygon-digital-carbon/src/KlimaAggregator.ts

```diff
@@ -1,215 +1,206 @@
-  if (event.params.carbonPool != ZERO_ADDRESS) retire.pool = event.params.carbonPool
+  let onchainIndex = getOnchainRetirementIndex(beneficiaryAddress)
 
-  retire.source = 'KLIMA'
-  retire.beneficiaryAddress = event.params.beneficiaryAddress
-  retire.beneficiaryName = event.params.beneficiaryString
-  retire.retiringAddress = event.params.retiringAddress
-  retire.retirementMessage = event.params.retirementMessage
+  let diff: i32 = onchainIndex.minus(BigInt.fromI32(sender.previousRetirementIndex)).toI32()
+  // @note we could technically just use the previousRetirementIndex here versus retrieve the onchain index
+  // this method retains a link to the onchain index for syncing purposes
+  // if there is no reason to keep that link to the onchain index we can just assign index as previousRetirementIndex and increment
 
-  retire.save()
+  let adjustedIndex = onchainIndex.minus(BigInt.fromI32(diff))
 
-  saveKlimaRetire(
-    event.params.beneficiaryAddress,
+  const klimaRetire = saveKlimaRetire(
+    beneficiaryAddress,
     retire.id,
-    index,
-    event.params.retiredAmount.div(BigInt.fromI32(100)),
+    adjustedIndex,
+    retiredAmount.div(BigInt.fromI32(100)), // hard-coded 1% fee
     false
   )
+
+  // Generate daily retirement data if needed
+  if (klimaRetire !== null) {
+    const dailyRetirement = generateDailyKlimaRetirement(klimaRetire)
+    if (dailyRetirement !== null) {
+      dailyRetirement.save()
+    }
+  }
+
+  // Update protocol metrics if needed
+  if (retire.pool !== null && Address.fromBytes(retire.pool as Bytes) != ZERO_ADDRESS) {
+    updateKlimaRetirementProtocolMetrics(retire.pool as Bytes, timestamp, retiredAmount)
+  }
 }
 
-export function handleToucanRetired(event: ToucanRetired): void {
+export function handleMossRetired(event: MossRetired): void {
   // Ignore zero value retirements
   if (event.params.retiredAmount == ZERO_BI) return
-  let network = dataSource.network()
-
-  let retirementsContractAddress = getRetirementsContractAddress(network)
-  let klimaRetirements = KlimaCarbonRetirements.bind(retirementsContractAddress)
-
-  let index = klimaRetirements.retirements(event.params.beneficiaryAddress).value0.minus(BigInt.fromI32(1))
 
   let sender = loadOrCreateAccount(event.transaction.from)
   loadOrCreateAccount(event.params.beneficiaryAddress)
   loadOrCreateAccount(event.params.retiringAddress)
 
-  let retire = loadRetire(sender.id.concatI32(sender.totalRetirements - 1))
+  processRetirement(
+    sender,
+    event.params.beneficiaryAddress,
+    event.params.beneficiaryString,
+    event.params.retiringAddress,
+    event.params.retirementMessage,
+    event.params.carbonPool,
+    event.params.retiredAmount,
+    event.block.timestamp
+  )
 
-  if (event.params.carbonPool != ZERO_ADDRESS) retire.pool = event.params.carbonPool
+  sender.previousRetirementIndex = sender.previousRetirementIndex + 1
+  log.info('sender.previousRetirementIndex: {}', [sender.previousRetirementIndex.toString()])
```

Comment:
Can we put that inside the `processRetirement` method ?

Are you keeping the logs on purpose?

Reply:
the logs were just for debugging the latest deployment. i moved the index increment ( it actually needed to be for beneficiary, not sender) inside `processRetirement` and removed the logs in the latest commit